### PR TITLE
Introduce container-image-podman extension

### LINKED ISSUE #40699: Create an extension for using podman to build container images
### Description

Currently we have Jib, Docker, OpenShift, and Buildpack container images. There isn't a way to use podman to build a container image and have it loaded into the local registry.

I'm proposing a new `container-image-podman` extension which would essentially mirror the `container-image-docker` extension and implement similar capabilities, as supported by podman.

### Implementation ideas

_No response_

### PR DESCRIPTION
**Pull request by edeandrea**
Introduce a new `container-image-podman` extension. See the conversation history on #40699 for the details that were discussed between myself and @geoand .

Essentially this refactors the current `container-image-docker` extension and adds a common layer underneath and pulls lots of common functionality up to that new common layer. Then it adds a new `container-image-podman` extension on top of the common layer.

- Closes: #40699

### CODE REVIEW

**Comment by edeandrea** [comment]
There were a bunch of refactorings where, because of the new layer (`ContainerRuntimeStatusBuildItem`) that `DockerStatusBuildItem.isDockerWorking()` was moved/renamed to `ContainerRuntimeStatusBuildItem.isContainerRuntimeWorking()`, so a lot of the changes in the various dev services are just that method name change.

File: extensions/container-image/container-image-docker-common/pom.xml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>quarkus-container-image-docker-common-parent</artifactId>
+    <name>Quarkus - Container Image - Docker - Common</name>
+    <packaging>pom</packaging>
+    <modules>
+        <module>deployment</module>
+        <module>runtime</module>
+        <module>container-image-docker</module>
+        <module>container-image-podman</module>
```
Comment (geoand) [inline extensions/container-image/container-image-docker-common/pom.xml:19]: What's the rationale for having these **under**  common? It's not something I would expect to see if I opened the directory for the first time

File: extensions/container-image/container-image-docker-common/runtime/src/main/resources/META-INF/quarkus-extension.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+name: "Container Image - Docker common"
+metadata:
+  keywords:
+    - "container"
+    - "image"
+    - "docker"
+  categories:
+    - "cloud"
+  status: "preview"
+  unlisted: true
```
Comment (geoand) [inline extensions/container-image/container-image-docker-common/runtime/src/main/resources/META-INF/quarkus-extension.yaml:12]: 👍🏼

File: core/deployment/src/main/java/io/quarkus/deployment/util/ContainerRuntimeUtil.java [hunk 3]
```diff
@@ -70,47 +80,58 @@ public static ContainerRuntime detectContainerRuntime(boolean required) {
...
-                podmanVersionOutput = getVersionOutputFor(ContainerRuntime.PODMAN);
-                podmanAvailable = PODMAN_PATTERN.matcher(podmanVersionOutput).matches();
-                if (podmanAvailable) {
-                    return ContainerRuntime.PODMAN;
-                }
+            var runtime = runtimesToCheck.stream()
+                    .filter(containerRuntime -> CONTAINER_EXECUTABLE.trim()
+                            .equalsIgnoreCase(containerRuntime.getExecutableName()))
+                    .findFirst()
+                    .filter(r -> {
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkus/deployment/util/ContainerRuntimeUtil.java:96]: I am always surprised when people write this kind of code instead of a simple for each + if statement. Furthermore, I don't understand why we need this and the for loop starting at line `114`

File: extensions/container-image/container-image-docker-common/deployment/src/main/java/io/quarkus/container/image/docker/common/deployment/RedHatOpenJDKRuntimeBaseProvider.java [hunk 4]
```diff
@@ -22,7 +21,7 @@ public Optional<DockerFileBaseInformation> determine(Path dockerFile) {
             if (fromOpt.isPresent()) {
                 String fromLine = fromOpt.get();
                 String baseImage = fromLine.substring(4).trim();
-                Pattern pattern = Pattern.compile(".*ubi8/openjdk-(\\w+)-runtime.*");
+                Pattern pattern = Pattern.compile(".*ubi[89]/openjdk-(\\w+)-runtime.*");
```
Comment (geoand) [inline extensions/container-image/container-image-docker-common/deployment/src/main/java/io/quarkus/container/image/docker/common/deployment/RedHatOpenJDKRuntimeBaseProvider.java:24]: These kinds of behavior changes should be separate commits

File: extensions/container-image/container-image-docker-common/pom.xml (hunk 1 above)
Comment (geoand) [reply extensions/container-image/container-image-docker-common/pom.xml:19]: `container-image-podman` can be on the same level as the others

File: core/deployment/src/main/java/io/quarkus/deployment/util/ContainerRuntimeUtil.java (hunk 3 above)
Comment (edeandrea) [reply core/deployment/src/main/java/io/quarkus/deployment/util/ContainerRuntimeUtil.java:96]: > I am always surprised when people write this kind of code instead of a simple for each + if statement

To me this code is far more readable and shows its intention better than the code that was there beforehand. If you prefer it the other way I can re-write it.

> Furthermore, I don't understand why we need this and the for loop starting at line `114`

It was already there before. I didn't really change the structure of the code.

https://github.com/quarkusio/quarkus/blob/1a584779b81dc22741291285a04e0a94c1740f3a/core/deployment/src/main/java/io/quarkus/deployment/util/ContainerRuntimeUtil.java#L82-L99

File: extensions/container-image/container-image-docker-common/deployment/src/main/java/io/quarkus/container/image/docker/common/deployment/RedHatOpenJDKRuntimeBaseProvider.java (hunk 4 above)
Comment (edeandrea) [reply extensions/container-image/container-image-docker-common/deployment/src/main/java/io/quarkus/container/image/docker/common/deployment/RedHatOpenJDKRuntimeBaseProvider.java:24]: Sure. I can revert this and move to a separate commit.

File: extensions/container-image/container-image-docker-common/pom.xml (hunk 1 above)
Comment (edeandrea) [reply extensions/container-image/container-image-docker-common/pom.xml:19]: Just so I'm clear, you mean `container-image-podman`, `container-image-docker`, and `container-image-docker-common` should all be on the same level as the others, and `container-image-podman` and `container-image-docker` should include `container-image-docker-common` as a dependency?

Comment (geoand) [reply extensions/container-image/container-image-docker-common/pom.xml:19]: yup

File: core/deployment/src/main/java/io/quarkus/deployment/util/ContainerRuntimeUtil.java (hunk 3 above)
Comment (edeandrea) [reply core/deployment/src/main/java/io/quarkus/deployment/util/ContainerRuntimeUtil.java:96]: The reason I had to even touch this code is because if I have the `container-image-podman` extension present but I have both docker & podman installed on my machine, it was always finding docker first and trying to build with docker.

This change makes it so the caller can control the search order (if they want to - its backwards compatible to other callers that don't care).

Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/util/ContainerRuntimeUtil.java:96]: > It was already there before. I didn't really change the structure of the code.

Okay, I see that

File: extensions/container-image/container-image-docker-common/pom.xml (hunk 1 above)
Comment (edeandrea) [reply extensions/container-image/container-image-docker-common/pom.xml:19]: Will do. I'll make this change. Sorry about that! I just assumed when one extension extends another it implies hierarchical.

Comment (geoand) [reply extensions/container-image/container-image-docker-common/pom.xml:19]: np :)

File: extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java
Comment (edeandrea) [inline extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java:1]: No idea why this file was changed. Maybe IntelliJ reformatted it?

File: extensions/container-image/container-image-docker-common/pom.xml (hunk 1 above)
Comment (edeandrea) [reply extensions/container-image/container-image-docker-common/pom.xml]: Done.

File: extensions/container-image/container-image-docker-common/deployment/src/main/java/io/quarkus/container/image/docker/common/deployment/RedHatOpenJDKRuntimeBaseProvider.java (hunk 4 above)
Comment (edeandrea) [reply extensions/container-image/container-image-docker-common/deployment/src/main/java/io/quarkus/container/image/docker/common/deployment/RedHatOpenJDKRuntimeBaseProvider.java]: Done

File: extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java
Comment (edeandrea) [reply extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java:1]: I did run the formatter goal on the project. Maybe this file was out of format? IDK. I didn't change anything in here.

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/ContainerRuntimeStatusBuildItem.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    protected ContainerRuntimeStatusBuildItem(IsContainerRuntimeWorking isContainerRuntimeWorking) {
+        this.isContainerRuntimeWorking = isContainerRuntimeWorking;
+    }
+
+    public boolean isContainerRuntimeAvailable() {
+        synchronized (this) {
+            if (cachedStatus == null) {
+                cachedStatus = isContainerRuntimeWorking.getAsBoolean();
+            }
+        }
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkus/deployment/builditem/ContainerRuntimeStatusBuildItem.java:19]: Not too important, but this could be improved like so:

```suggestion
        if (cachedStatus == null) {
          synchronized (this) {
            if (cachedStatus == null) {
                cachedStatus = isContainerRuntimeWorking.getAsBoolean();
            }
          }
        }
```

Comment (edeandrea) [reply core/deployment/src/main/java/io/quarkus/deployment/builditem/ContainerRuntimeStatusBuildItem.java:19]: Good catch! I'll make that change this morning.

Originally the entire method had the `synchronized` keyword on it, so I think this is a big improvement. No need to synchronize the whole method.

**Comment by edeandrea** [comment]
Not sure why the documentation CI is failing. I didn't change any of the docs since the last build which succeeded. I did rebase from main, so maybe something there broke it?

**Comment by geoand** [comment]
#40775 was merged which hopefully fixes the doc issue

**Comment by edeandrea** [comment]
> #40775 was merged which hopefully fixes the doc issue

I just rebased from main again to pull this in.

**Review by geoand — APPROVED** [review]
LGTM, thanks!
